backend/firebase/firebase.tsx

```diff
@@ -34,4 +33,32 @@ const providers = {
   twitter: new TwitterAuthProvider(),
 };
 
-export { auth, db, app, analytics };
+const signInWithGoogle = async () => {
+  try {
+    const res = await signInWithPopup(auth, providers.google);
+    const user = res.user;
+    const q = query(collection(db, "User + Admin + Volunteer"), where("uid", "==", user.uid));
+    const docs = await getDocs(q);
+
+    if (docs.docs.length === 0) {
+      await addDoc(collection(db, "User + Admin + Volunteer"), {
+        user_id: user.uid,
+        name: user.displayName,
+        authProvider: "google",
+        email: user.email,
+      });
+    }
+  }
+  catch (err) {
+    console.error(err);
+    if (err instanceof Error) {
+      alert("An error occurred while signing in with Google: " + err.message);
+    }
+  }
+};
+
+const logout = () => {
+  signOut(auth);
+};
```

Comment:
Some syntactic sugar:

```suggestion
const logout = () => signOut(auth);
```